sys/include/zptr.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2017 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_util_zptr Pointer Compression
+ * @ingroup     sys
+ * @brief       Provides 32bit -> 16bit pointer compression
+ *
+ * On many platforms, some pointers may have to be aligned, e.g., to 4 byte
+ * bounderies.
+ * On 32bit platforms, that makes it possible to store all possible aligned
+ * 32bit pointers in a 16bit value as long as the total memory is small (e.g.,
+ * with 4 byte alignment, all pointers within 256kb RAM can be represented by a
+ * 16bit value). This can save memory, at the cost of some instructions for
+ * compression/decompression.
+ *
+ * In order to use pointer compression, ZPTR_BASE needs to be defined to a (4
+ * byte aligned) base address.
+ *
+ * A printf format macro (PRIzptr) is provided.
+ *
+ * You can then use zptr_t instead of a pointer type, using the supplied functions to compress / decompress,
+ * e.g.,
+ *
+ *     void func(void *ptr) {
+ *         printf("%"PRIzptr"\n", ptr);
+ *         ...
+ *         free(ptr);
+ * }
+ *
+ * ... would become
+ *
+ *     void func(zptr_t zptr);
+ *         printf("%"PRIzptr"\n", zptr);
+ *         ...
+ *         free(zptrd(zptr));
+ *     }
+ *
+ *
+ * If ZPTR_BASE is unset, @ref zptr_t / @ref zptrc() / @ref zptrd() will
+ * transparently and without overhead compile to normal (uncompressed) pointer
+ * operations.
+ *
+ * @{
+ *
+ * @file
+ * @brief       32bit -> 16bit pointer compression implementation
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ */
+
+#ifndef ZPTR_H
+#define ZPTR_H
+
+#include <assert.h>
+#include <stdint.h>
+#include <inttypes.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if ZPTR_BASE || defined(DOXYGEN)
+
+/**
+ * @brief zptr type definition
+ */
+typedef uint16_t zptr_t;
+
+/**
+ * @brief zptr printf format definition
+ */
+#define PRIzptr PRIu16
+
+/**
+ * @brief Compress a pointer (if possible)
+ *
+ * Substracts ZPTR_BASE, then right-shifts @p pointer by two.
+ *
+ * @param[in]   pointer     pointer to compress
+ * @returns     compressed pointer
+ */
+static inline zptr_t zptrc(void *pointer)
+{
+    assert(!(pointer & 0x3));
```

Comment:
```suggestion
    assert(!(pointer & 0x3) && !(pointer & 0xFFFC0000));
```

Reply:
Ah, the aim is to ensure this is within range? Then it needs to be sth like `pointer & 0x3 && (intptr_t)pointer >= ZPTR_BASE && (intptr_t) pointer <= ZPTR_BASE + 2**18`.
Il factor that into a function...